app/controllers/users/passwords_controller.rb

```diff
@@ -6,9 +6,15 @@ class Users::PasswordsController < Devise::PasswordsController
   layout "session"
 
   def create
-    super do |user|
-      if user.nil? || user.new_record?
-        flash.now[:notice] = ts("We couldn't find an account with that email address or username. Please try again?")
+    user_id = User.find_for_authentication(login: resource_params[:login])&.id
```

Comment:
You can do `User.find_for_authentication(resource_params)`.

Reply:
Hmm. I'll try it again, but it didn't seem to ever find the user when I did that -- the test would fail here:

```    And I fill in "Email address or user name" with "target"                                                   # features/step_definitions/web_steps.rb:52
    And I press "Reset Password"                                                                               # features/step_definitions/web_steps.rb:40
    Then I should be on the home page                                                                          # features/step_definitions/web_steps.rb:268
      expected: "/"
           got: "/users/login" (using ==) (RSpec::Expectations::ExpectationNotMetError)
      ./features/step_definitions/web_steps.rb:271:in `/^(?:|I )should be on (.+)$/'
      features/users/authenticate_users.feature:238:in `Then I should be on the home page'
      ```

Reply:
Now that I have straightened out the password column issue, ahem, I'm getting a more helpful error, which happens in every scenario we press the Reset Password button:

```
      unable to convert unpermitted parameters to hash (ActionController::UnfilteredParameters)
      ./app/models/user.rb:203:in `find_first_by_auth_conditions'
      ./app/controllers/users/passwords_controller.rb:9:in `create'
      ./features/step_definitions/web_steps.rb:42:in `block (2 levels) in <top (required)>'
      ./features/step_definitions/web_steps.rb:6:in `with_scope'
      ./features/step_definitions/web_steps.rb:41:in `/^(?:|I )press "([^"]*)"(?: within "([^"]*)")?$/'
      features/users/authenticate_users.feature:237:in `And I press "Reset Password"'
```

The referenced line is 203: https://github.com/otwcode/otwarchive/blob/c2f27c95463958b1fd184670616364ba9592029c/app/models/user.rb#L202-L203

But the `resource_params` are definitely `{"login"=>"sarken"}`

But testing manually with the code I have (`User.find_for_authentication(login: resource_params[:login])&.id`) is also erroring now, even though the tests are passing. :thinking:

Reply:
Okay, I _did_ figure out why my manual tests were erroring: I needed to restart after changing the name of the config  variable. That's not embarrassing at all.

But what I can't figure out is how to sweet talk Devise into permitting these parameters. The [documentation](https://github.com/heartcombo/devise#strong-parameters) made it seem like you could only permit additional parameters for three particular actions from the Registrations or Sessions controllers, but the [source](https://github.com/heartcombo/devise/blob/45b831c4ea5a35914037bd27fe88b76d7b3683a4/lib/devise/parameter_sanitizer.rb#L82-L87) made it seem like I could specify any action for any controller... but no luck with either of these as a `before_action` in the Passwords controller:

```
  def configure_permitted_parameters
    devise_parameter_sanitizer.permit(:create) do |user|
      user.permit(:login)
    end
  end

  def configure_permitted_parameters
    devise_parameter_sanitizer.permit(:create, keys: [:login])
  end
  ```

I even tried using sign_up, sign_in, and account_update just for the hell of it, in case they were the only allowed options.

Reply:
For the record (1):
```rb
before_action :configure_permitted_parameters, only: [:create]
def configure_permitted_parameters
  devise_parameter_sanitizer.permit(:create, keys: [:login])
end
```
This would work if you also call:
```rb
User.find_for_authentication(devise_parameter_sanitizer.sanitize(:create))
```

For the record (2): with password resets, Devise ends up permitting all parameters, and does [extra filtering to get the `:login` key](https://github.com/heartcombo/devise/blob/45b831c4ea5a35914037bd27fe88b76d7b3683a4/lib/devise/models/authenticatable.rb#L290-L297). So no different from `login: resource_params[:login]`.

That said, there's a nicer way we can write it:
```rb
User.find_for_authentication(resource_params.permit(:login))
```

Reply:
Ah, thank you! I think that makes sense to me now. For some reason, I was thinking `configure_permitted_parameters` alone would magically alter how `find_for_authentication` treated `resource_params`. But that was silly -- it's not like defining the permitted parameters works that way in any other controller; we define the parameters and then reference that method when passing it along to the model.